src/line_launcher/mod.rs

```diff
@@ -304,193 +504,208 @@ impl LineLauncher {
-                            .send(ScheduleNoteOnMessage {
-                                time: SystemTime::now()
-                                    + duration_between_sixteenth_notes.mul_f64(
-                                        (1.0 + ((0.5
-                                            - ahead_or_behind_the_beat_ratio.unwrap_or(0.5))
-                                            * 2.0)
-                                            + random_ratio_of_duration_between_sixteenth_notes)
-                                            .max(0.0),
-                                    ),
-                                note_index: next_note_index,
-                            })
-                            .unwrap();
-
-                        *playing_state = PlayingState::Playing {
-                            line_index,
-                            next_note_index: next_note_index + 1,
-                            pitch_offset,
-                            next_note_off_index,
-                        };
-                    } else {
-                        panic!(
-                            "Couldn't schedule note on because no duration between sixteenth notes"
-                        );
-                    }
+                let second_note_in_new_outside_key: u8 = line.notes[1]
+                    .note
+                    .step(new_outside_of_the_key_offset)
+                    .unwrap()
+                    .into();
+                let (is_within_a_half_step, second_octave_adjustment) = get_is_within_a_half_step(
+                    second_note_in_new_outside_key,
+                    last_two_planned_notes.1,
+                );
+                if !is_within_a_half_step {
+                    return false;
                 }
-            }
-            _ => {
-                panic!(
-                    "Called possibly_schedule_note_on() while not playing: {:?}",
-                    playing_state
+                if first_octave_adjustment != second_octave_adjustment {
+                    return false;
+                }
+                octave_adjustment = first_octave_adjustment;
+                true
+            });
+            if let Some(new_line_index) = new_line_index {
+                let new_outside_of_the_key_offset_octave_adjusted =
+                    new_outside_of_the_key_offset + octave_adjustment;
+                debug!(
+                    "found line index: {}, octave-adjusted outside_of_the_key_offset: {}",
+                    new_line_index, new_outside_of_the_key_offset_octave_adjusted
+                );
+                return (
+                    new_line_index,
+                    new_outside_of_the_key_offset_octave_adjusted,
                 );
             }
         }
     }
 }
+
+fn get_is_within_a_half_step(note_a: u8, note_b: u8) -> (bool, i8) {
+    let raw_difference = note_a as i8 - note_b as i8;
+    let difference_within_same_octave = raw_difference.rem_euclid(12);
```

Comment:
I don't follow from the docs on `rem_euclid` when you'd use this over `%` and vice versa.